[bugfix](font-weight): Fix rendering of font-weight and other

d-desiatkin opened the pull request:
[bugfix](font-weight): Fix rendering of font-weight and other propperties related to FontDescriptor
---

<!-- Please describe your changes on the following line: -->

<details>
<summary>Wrongly identified cause. Initially I correctly added font file support, and that allowed code bellow to solve the issue:</summary>
<p>
After font redesign servo have an idea of creation Font object that will share some fields. Like font-weight, font-width, and other fields in font descriptor shouldn't have unique values within the font (their modification shouldn't happen through new Font object creation). However Font matching (check that symbol is supported by Font) process was not adapted for this new change. On Font resolution we will create a lot of new Fonts. Basically any change in FontDescriptor object will create new unique Font object.

Also this idea contradicts with current realization of TextRun / GlyphRun caching strategy. This cache is now stored inside Font class, and such class should have unique values of FontDescriptor field.

Lets decompose `TextRun::segment_text_by_font` method:
1) We get a handle for entry of global cache of font files (cache itself is in another process) `FontGroup`
`let font_group = font_context.font_group(self.parent_style.clone_font());` 
This font group currently unified by Font pt size and style of parent element (font-size should be computed during layout. This is the reason why it separated and couldn't be taken from style). If such combination of font-size and parent style was not observed before, new font_group object will be created and placed into `FontContext` (global font cache).
2) We start to resolve individual symbols in `IFC` accumulated text string with this `FontGroup`. If we met a symbol that can be resolved with some Font that already exists in `FontGroup` it's reference will be returned, otherwise new Font will be created within the FontGroup that is able to resolve the symbol and it's Reference will be returned. Result of this operation is `Font` object to use for shaping and rendering of current symbol
3) We take decision whether we can extend current `TextRunSegment` that is under construction (namings is really bad here, current `TextRun` should be renamed to `CSSTextSequence`, current `TextRunSegment` should be renamed to `TextRun`). `TextRunSegment` should have 4 properties in common: 
- language
- script
- font
- bidi direction
If all this properties for current symbol corresponds to same properties from different symbol we add current symbol to  
`TextRunSegment` otherwise we must start new `TextRunSegment`. Right now font check is implemented poorly on this step. "TextRunSegment::update_if_compatible" is responsible for that check let's decompose this function

---

`TextRunSegment::update_if_compatible` currently check bidi_level correspondance and return false if it is different.
Later it checks font for equality. This check is implemented by comparison of key of font in global FontContext and size of the Font. For me it seems that author assumed that Font::key() uniquely identify fields of the font file. But this is simply not true. Check codeblock that will generate this Font::key() bellow

```rust
pub(crate) fn create_font_instance_key(&self, font: &Font) -> FontInstanceKey {
        match font.template.identifier() {
            FontIdentifier::Local(_) => self.system_font_service_proxy.get_system_font_instance(
                font.template.identifier(),
                font.descriptor.pt_size,
                font.webrender_font_instance_flags(),
            ),
            FontIdentifier::Web(_) => self.create_web_font_instance(
                font.template.clone(),
                font.descriptor.pt_size,
                font.webrender_font_instance_flags(),
            ),
        }
    }
```

As you can see we use `font.template.identifier()` this object contains descriptor of the font inside. But this descriptor is platform specific and contains only default descriptor of the font generated by platform font manager. CSS styling will not take any efect on font descriptor generated by platform.  That is why we need to add font descriptor information in this key somehow, right now we add only  `font.descriptor.pt_size,`. Also we add `font.webrender_font_instance_flags(),` Some rendering specific information, but it is not that important right now.

Insufficient check leads to the following bug. We resolve correct font during symbol resolution, so we allways create needed fonts. But we will never use part of them due to the fact that 2 fonts can have `Font::key()` and simmilar `font-sizes`. So we are never extract them from storage to do actual work.

---

So now Reviewer should have clear picture of the problem. We have 2 Solutions. 
 - First change `Font::key()` in global `FontContext` and make it include all information about font_descriptor and not only the size.
 - Second add aditional check inside `TextRunSegment::update_if_compatible` function that will properly compare all fields of FontDescriptor. That solution will be platform independent and less intrusive.

As you see this PR contains code for second approach. We compute Hash across the fields of FontDescriptor and compare it to make decision whether we can add current symbol to TextRunSegment under construction.
</p>
</details>

We need to change the way we create variations for font-templates on OpenHarmony platform. 
`components\fonts\platform\freetype\ohos\font_list.rs`

In case servo wants to rely on self written font manager we must parse OpenHarmony analogue of font config on the following path '/etc/fontconfig.json'.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors

These potentially related to #32033 

<!-- Either: -->
- [ ] This PR fixes #35784. Test for that PR could be found in the issue

This PR is related to the followng issues:
- #32723
- #32644

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

Loirooriol commented:
Yeah, no testcase provided and no WPT passing, so it's hard to know what this is improving.
Also, clippy is complaining: https://github.com/servo/servo/actions/runs/13564140590/job/37913359746?pr=35696

d-desiatkin commented:
> Yeah, no testcase provided and no WPT passing, so it's hard to know what this is improving. Also, clippy is complaining: https://github.com/servo/servo/actions/runs/13564140590/job/37913359746?pr=35696

There was an issue indeed. I was in a hurry yesterday. I restored font-size check cause it needed. You can check updated description above. Also I have attached my minimal test case.

Will also add link to affected WPT tests.

Loirooriol commented:
I don't see any difference in the testcase before vs after your changes.

Maybe the problem was fixed by #35705?

d-desiatkin commented:
Yeah. For me it seems more like font fallback problem than specific issue with font-weight.
I believe font-templates are created in a wrong way on OpenHarmony comaped to other platforms. That is why FontGroup::find_by_codepoint behaves differently. 

We have 2 potentioal problems:
 - choose wrong FontGroup
 - resolution of Font within one FontGroup.

Do you have any specific document that shows Font module design? I want to compose one to enhance my understanding of the issue.

d-desiatkin commented:
> I don't see any difference in the testcase before vs after your changes.
> 
> Maybe the problem was fixed by #35705?

Previously that code fixed inconsistencies in style. Now it seems fallback strategy is more suitable indeed

Loirooriol commented:
I don't understand the current situation. If there is still some broken testcase, please post it. Otherwise I'm not sure what you are trying to achieve.

d-desiatkin commented:
Yeah. I will create an issue

d-desiatkin commented:
https://github.com/servo/servo/issues/35784

jschwe commented on components/fonts/platform/freetype/ohos/json.rs:
```diff
@@ -0,0 +1,487 @@
+/// Representation of OpenHarmony /etc/fontconfig.json
+/// in Rust structure
+#[derive(Debug)]
+pub struct FontconfigOHOS {
+    pub fontdir: String,
+    pub generic: Vec<GenericFontFamilyOHOS>,
+    pub fallback: Vec<(String, Vec<FallbackEntry>)>,
+    pub font_file_map: HashMap<String, String>,
+}
```
Any particular reason why you don't derive Deserialize?

mrobinson commented on components/layout_2020/flow/inline/mod.rs:
```diff
@@ -172,8 +172,22 @@ pub(crate) struct InlineFormattingContext {
 /// A collection of data used to cache [`FontMetrics`] in the [`InlineFormattingContext`]
 #[derive(Debug)]
 pub(crate) struct FontKeyAndMetrics {
+    // Font instance key allows us to get font from system
+    // However this font doesn't contain information that
+    // stored in font_descriptor, only about template.descriptor.
+    // Nevertheless it seems that
+    // created keys would be unique for the font files.
```
What information is missing exactly? I think the bug here is that `FontInstanceKey` isn't identifying the font uniquely, but this change just papers around the issue.

mrobinson commented on components/fonts/font.rs:
```diff
@@ -609,6 +609,7 @@ impl FontGroup {
         let should_look_for_small_caps = self.descriptor.variant == font_variant_caps::T::SmallCaps &&
             options.character.is_ascii_lowercase();
         let font_or_synthesized_small_caps = |font: FontRef| {
+            // log::warn!("font descriptor of codepoint {:?}", font.descriptor);
```
Please remove this.

mrobinson commented on components/fonts/lib.rs:
```diff
@@ -52,6 +52,7 @@ impl AsRef<[u8]> for FontData {
 /// Whether or not font fallback selection prefers the emoji or text representation
 /// of a character. If `None` then either presentation is acceptable.
 #[derive(Clone, Copy, Debug, PartialEq)]
+#[repr(u8)]
```
Why is this necessary?

mrobinson commented on components/fonts/platform/freetype/ohos/font_list.rs:
```diff
@@ -14,15 +15,18 @@ use style::values::computed::font::GenericFontFamily;
 use style::values::computed::{
     FontStretch as StyleFontStretch, FontStyle as StyleFontStyle, FontWeight as StyleFontWeight,
 };
+// use style::values::specified::font::{MIN_FONT_WEIGHT, MAX_FONT_WEIGHT};
```
Please remove this.

mrobinson commented on components/fonts/platform/freetype/ohos/font_list.rs:
```diff
@@ -14,15 +15,18 @@ use style::values::computed::font::GenericFontFamily;
 };
 
 static FONT_LIST: LazyLock<FontList> = LazyLock::new(FontList::new);
+// static FALLBACK_ASSOCIATIONS: LazyLock<FallbackAssociations> = LazyLock::new(generate_default_fallback_associations);
```
Please remove commented out code.

mrobinson commented on components/fonts/platform/freetype/ohos/json.rs:
```diff
@@ -0,0 +1,814 @@
+
+static FONT_CONFIG_PATH: &str = "/etc/fontconfig.json";
+
+// This file contains custom functions for json parsing to avoid
+// blind reliance on serde_json::from_*. Even if user make some mistake in
+// fontconfig we want to inform him about the problem, but rest of usefull information
+// should be preserved and used in servo engine.
```
It feels like maybe this is a bridge too far. It should be fine to not implement everything manually below and just report and error if the file cannot be parsed. The amount of parsing code below feels wrong somehow.

d-desiatkin commented:
```bash
Generic font families from config:
FontFamily { name: "HarmonyOS Sans", fonts: [Font { filepath: "/system/fonts/HarmonyOS_Sans.ttf", weight: Some(100), style: None, width: Normal }, Font { filepath: "/system/fonts/HarmonyOS_Sans.ttf", weight: Some(300), style: None, width: Normal }, Font { filepath: "/system/fonts/HarmonyOS_Sans.ttf", weight: Some(400), style: None, width: Normal }, Font { filepath: "/system/fonts/HarmonyOS_Sans.ttf", weight: Some(500), style: None, width: Normal }, Font { filepath: "/system/fonts/HarmonyOS_Sans.ttf", weight: Some(700), style: None, width: Normal }, Font { filepath: "/system/fonts/HarmonyOS_Sans.ttf", weight: Some(900), style: None, width: Normal }, Font { filepath: "HarmonyOS Sans Naskh Arabic UI", weight: Some(100), style: None, width: Normal }, Font { filepath: "HarmonyOS Sans Naskh Arabic UI", weight: Some(300), style: None, width: Normal }, Font { filepath: "HarmonyOS Sans Naskh Arabic UI", weight: Some(400), style: None, width: Normal }, Font { filepath: "HarmonyOS Sans Naskh Arabic UI", weight: Some(500), style: None, width: Normal }, Font { filepath: "HarmonyOS Sans Naskh Arabic UI", weight: Some(700), style: None, width: Normal }, Font { filepath: "HarmonyOS Sans Naskh Arabic UI", weight: Some(900), style: None, width: Normal }, Font { filepath: "HarmonyOS Sans SC", weight: Some(100), style: None, width: Normal }, Font { filepath: "HarmonyOS Sans SC", weight: Some(300), style: None, width: Normal }, Font { filepath: "HarmonyOS Sans SC", weight: Some(400), style: None, width: Normal }, Font { filepath: "HarmonyOS Sans SC", weight: Some(500), style: None, width: Normal }, Font { filepath: "HarmonyOS Sans SC", weight: Some(600), style: None, width: Normal }, Font { filepath: "HarmonyOS Sans SC", weight: Some(700), style: None, width: Normal }, Font { filepath: "HarmonyOS Sans SC", weight: Some(900), style: None, width: Normal }, Font { filepath: "HarmonyOS Sans TC", weight: Some(100), style: None, width: Normal }, Font { filepath: "HarmonyOS Sans TC", weight: Some(300), style: None, width: Normal }, Font { filepath: "HarmonyOS Sans TC", weight: Some(400), style: None, width: Normal }, Font { filepath: "HarmonyOS Sans TC", weight: Some(500), style: None, width: Normal }, Font { filepath: "HarmonyOS Sans TC", weight: Some(700), style: None, width: Normal }, Font { filepath: "HarmonyOS Sans TC", weight: Some(900), style: None, width: Normal }] }
FontFamily { name: "HarmonyOS Sans Condensed", fonts: [Font { filepath: "/system/fonts/HarmonyOS_Sans_Condensed.ttf", weight: Some(100), style: None, width: Normal }, Font { filepath: "/system/fonts/HarmonyOS_Sans_Condensed.ttf", weight: Some(300), style: None, width: Normal }, Font { filepath: "/system/fonts/HarmonyOS_Sans_Condensed.ttf", weight: Some(400), style: None, width: Normal }, Font { filepath: "/system/fonts/HarmonyOS_Sans_Condensed.ttf", weight: Some(500), style: None, width: Normal }, Font { filepath: "/system/fonts/HarmonyOS_Sans_Condensed.ttf", weight: Some(700), style: None, width: Normal }, Font { filepath: "/system/fonts/HarmonyOS_Sans_Condensed.ttf", weight: Some(900), style: None, width: Normal }] }
FontFamily { name: "HarmonyOS Sans Digit", fonts: [Font { filepath: "/system/fonts/HarmonyOS_Sans_Digit.ttf", weight: None, style: None, width: Normal }] }
FontFamily { name: "Noto Serif", fonts: [Font { filepath: "/system/fonts/NotoSerif[wdth,wght].ttf", weight: None, style: None, width: Normal }, Font { filepath: "Noto Serif Ahom Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Serif Tibetan Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Serif Dogra Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Serif Yezidi Regular", weight: None, style: None, width: Normal }] }
FontFamily { name: "Noto Sans Mono", fonts: [Font { filepath: "/system/fonts/NotoSansMono[wdth,wght].ttf", weight: None, style: None, width: Normal }] }
FontFamily { name: "HMOS Color Emoji", fonts: [Font { filepath: "/system/fonts/HMOSColorEmojiCompat.ttf", weight: None, style: None, width: Normal }, Font { filepath: "HMOS Color Emoji", weight: None, style: None, width: Normal }, Font { filepath: "HMOS Color Emoji Flags", weight: None, style: None, width: Normal }] }
FontFamily { name: "HMOS Color Emoji Flags", fonts: [Font { filepath: "/system/fonts/HMOSColorEmojiFlags.ttf", weight: None, style: None, width: Normal }, Font { filepath: "HMOS Color Emoji Flags", weight: None, style: None, width: Normal }] }
FontFamily { name: "Noto Sans", fonts: [Font { filepath: "Noto Sans Adlam Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Anatolian Hieroglyphs Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Armenian Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Avestan Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Balinese Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Bamum Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Bassa Vah Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Batak Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Bengali UI Bold", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Bengali UI Medium", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Bengali UI Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Bengali UI SemiBold", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Bengali Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Bhaiksuki Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Brahmi Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Buginese Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Buhid Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans CJK JP", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Canadian Aboriginal Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Carian Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Chakma Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Cham Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Cherokee Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Coptic Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Cuneiform Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Cypriot Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Deseret Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Devanagari UI Bold", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Devanagari UI Medium", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Devanagari UI Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Devanagari UI SemiBold", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Devanagari Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Egyptian Hieroglyphs Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Elbasan Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Ethiopic Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Georgian Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Glagolitic Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Gothic Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Grantha Regular", weight: None, s
to Sans Gurmukhi UI Medium", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Gurmukhi UI Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Gurmukhi UI SemiBold", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Gurmukhi Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Hanifi Rohingya Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Hanunoo Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Hatran Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Hebrew Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Imperial Aramaic Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Inscriptional Pahlavi Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Inscriptional Parthian Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Javanese Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Kaithi Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Kannada UI Bold", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Kannada UI Medium", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Kannada UI Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Kannada UI SemiBold", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Kannada Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Kayah Li Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Kharoshthi Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Khmer Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Khojki Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Lao Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Lepcha Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Limbu Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Linear A Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Linear B Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Lisu Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Lycian Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Lydian Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Malayalam UI Bold", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Malayalam UI Medium", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Malayalam UI Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Malayalam UI SemiBold", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Malayalam Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Mandaic Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Manichaean Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Marchen Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Masaram Gondi Regular", weight: None, style: None, width: Normal }, Font { f
e: None, width: Normal }, Font { filepath: "Noto Sans Mono Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Mro Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Multani Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Myanmar Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans NKo Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Nabataean Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans New Tai Lue Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Newa Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Ogham Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Ol Chiki Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Old Italic Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Old North Arabian Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Old Permic Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Old Persian Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Old South Arabian Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Old Turkic Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Oriya Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Osage Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Osmanya Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Pahawh Hmong Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Palmyrene Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Pau Cin Hau Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Phags-Pa Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Phoenician Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Rejang Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Runic Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Samaritan Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Saurashtra Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Sharada Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Shavian Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Sinhala UI Bold", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Sinhala UI Medium", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Sinhala UI Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Sinhala UI SemiBold", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Sinhala Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Sora Sompeng Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Soyombo Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Sundanese Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Syloti Nagri Regular", weight: None, style: None, width:
a Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Tai Le Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Tai Tham Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Tai Viet Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Takri Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Tamil UI Bold", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Tamil UI Medium", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Tamil UI Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Tamil UI SemiBold", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Tamil Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Telugu UI Bold", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Telugu UI Medium", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Telugu UI Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Telugu UI SemiBold", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Telugu Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Thaana Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Thai Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Tifinagh Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Ugaritic Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Vai Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Wancho Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Warang Citi Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Yi Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Regular", weight: None, style: None, width: Normal }, Font { filepath: "Noto Sans Math Regular", weight: None, style: None, width: Normal }] }
```

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

d-desiatkin committed:
[feature](font_matching, ohos_fontconfig):

This commit modifies current font matching algorithm to conform with
CSS-fonts-4 specification. Also it introduce OpenHarmony fontconfig
mechanism.

Signed-off-by: Desiatkin Dmitrii <d.desyatkin@innopolis.university>

d-desiatkin committed:
[feature](platform_font): Linux/Ubuntu/Android add extraction of unicode
ranges information from font file.

Signed-off-by: Desiatkin Dmitrii <d.desyatkin@innopolis.university>

d-desiatkin committed:
[feature]: Attempt to add variation axes control

Signed-off-by: Desiatkin Dmitrii <d.desyatkin@innopolis.university>

xiaochengh commented on components/layout_2020/flow/inline/text_run.rs:
```diff
@@ -415,6 +415,8 @@ impl TextRun {
         font_cache: &mut Vec<FontKeyAndMetrics>,
         bidi_info: &BidiInfo,
     ) -> Vec<(TextRunSegment, FontRef)> {
+        // In propper language implementation script should also be provided from style
```
We can get it from the computed style via the `-x-lang` property.

xiaochengh reviewed (commented):
Right now the diff is pretty big. Better to split into smaller ones to help reviews.

I'm not sure if last resort font will be integrated into Servo directly. Could you split it to another branch?

d-desiatkin committed:
[feature](font-variations): Attempt to add font variations;
Problem: unable to get sfnt name of the axis

Signed-off-by: Desiatkin Dmitrii <d.desyatkin@innopolis.university>

xiaochengh reviewed (commented):
This PR is really big and hard to review.

Is it possible to:
1. Split it into smaller PRs, each with test cases to understand and verify what's going on
2. A design doc serving as an overall routemap

d-desiatkin committed:
[feature](variable-fonts): Add variable fonts

support for linux / android / OpenHarmony.
Currently support is incomplete. For web fonts everything will work as
intended. For local fonts there is still some bugs that must be fixed

Signed-off-by: Desiatkin Dmitrii <d.desyatkin@innopolis.university>
